tasks/task_wifi.c

```diff
@@ -36,23 +52,64 @@ static void task_wifi_scan_handler(retro_task_t *task)
    task_set_finished(task, true);
 }
 
-bool task_push_wifi_scan(retro_task_callback_t cb)
+static void task_wifi_enable_handler(retro_task_t *task)
 {
-   retro_task_t   *task = task_init();
+   driver_wifi_enable(true);
+
+   task_set_progress(task, 100);
+   task_set_finished(task, true);
+}
+
+static void task_wifi_disable_handler(retro_task_t *task)
+{
+   driver_wifi_enable(false);
 
+   task_set_progress(task, 100);
+   task_set_finished(task, true);
+}
+
+static void task_wifi_disconnect_handler(retro_task_t *task)
+{
+   wifi_network_info_t netinfo;
+
+   if (driver_wifi_connection_info(&netinfo))
+     driver_wifi_disconnect_ssid(&netinfo);
+
+   task_set_progress(task, 100);
+   task_set_finished(task, true);
+}
+
+static void task_wifi_connect_handler(retro_task_t *task)
+{
+   driver_wifi_connect_ssid(task->user_data);
+
+   task_set_progress(task, 100);
+   task_set_finished(task, true);
+}
+
+bool task_push_wifi_connect(retro_task_callback_t cb, void *netptr) {
+   char msg[128];
+   retro_task_t   *task = task_init();
+   wifi_network_info_t *netinfo = (wifi_network_info_t*)netptr;
    if (!task)
       return false;
+      
+   snprintf(msg, sizeof(msg), msg_hash_to_str(MSG_WIFI_CONNECTING_TO), netinfo->ssid);
 
-   /* blocking means no other task can run while this one is running,
-    * which is the default */
    task->type           = TASK_TYPE_BLOCKING;
    task->state          = NULL;
-   task->handler        = task_wifi_scan_handler;
+   task->handler        = task_wifi_connect_handler;
    task->callback       = cb;
-   task->title          = strdup(msg_hash_to_str(
-                           MSG_SCANNING_WIRELESS_NETWORKS));
-
+   task->title          = strdup(msg);
+   task->user_data      = malloc(sizeof(*netinfo));
```

Comment:
Memory leak! `task->user_data` isn't `free()`'d

Also, every `malloc()` and `calloc()` needs to be cast to the correct type, or certain builds will fail